app/uk/gov/hmrc/emailverification/repositories/PasscodeMongoRepository.scala

```diff
@@ -16,39 +16,119 @@
 
 package uk.gov.hmrc.emailverification.repositories
 
+import config.AppConfig
 import javax.inject.{Inject, Singleton}
 import org.joda.time.DateTime
 import org.joda.time.DateTimeZone.UTC
 import play.api.libs.json._
 import play.modules.reactivemongo.ReactiveMongoComponent
 import reactivemongo.api.indexes.{Index, IndexType}
-import reactivemongo.bson.{BSONDocument, BSONObjectID}
+import reactivemongo.bson.{BSONDateTime, BSONDocument, BSONInteger, BSONObjectID, BSONString}
 import reactivemongo.play.json.ImplicitBSONHandlers._
-import uk.gov.hmrc.emailverification.models.PasscodeDoc
+import uk.gov.hmrc.emailverification.models.EmailPasscodeException.MaxPasscodesAttemptsExceeded
+import uk.gov.hmrc.emailverification.models.{EmailPasscodeException, PasscodeDoc}
 import uk.gov.hmrc.http.logging.SessionId
 import uk.gov.hmrc.mongo.ReactiveRepository
 import uk.gov.hmrc.mongo.json.ReactiveMongoFormats
 
 import scala.concurrent.{ExecutionContext, Future}
 
 @Singleton
-class PasscodeMongoRepository @Inject()(mongoComponent: ReactiveMongoComponent)(implicit ec: ExecutionContext)
+class PasscodeMongoRepository @Inject()(mongoComponent: ReactiveMongoComponent, config:AppConfig)(implicit ec: ExecutionContext)
   extends ReactiveRepository[PasscodeDoc, BSONObjectID](
     collectionName = "passcode",
     mongo = mongoComponent.mongoConnector.db,
     domainFormat = PasscodeDoc.format,
     idFormat = ReactiveMongoFormats.objectIdFormats) {
 
-  def upsert(sessionId: SessionId, passcode: String, email: String, validityDurationMinutes: Int): Future[Unit] = {
+
+
+  /**
+   * add/update session record with passcode and email address to use.
+   * Returns the new/updated passcode doc with incremented emailAttempts count.
+   * Future fails with EmailPasscodeException.MaxEmailsExceeded if the emailAttempts count then exceeds maxEmailAttempts
+   * @see MaxEmailsExceeded
```

Comment:
you could create/use an ADT instead if you didn't want to throw/remember to catch exceptions

Reply:
or just an Either?!

Reply:
thanks guys :) I don't mind handling errors via failing futures but wondered if there was a preferred mechanism I _should_ be using, I've refactored it so the limit checking logic is up in the controller instead anyway, which kind of keeps the repository code clearer/cleaner I think.